### DISCUSSION TITLE
Decision.Evaluate not returning the trace in response

### DISCUSSION THREAD

**Discussion by aayushbhaglal** [gorules/zen#257] [Q&A]
Hello everyone,

I have been trying to use zen-go for my use case and have been facing an issue. I need the complete trace and result of each decision table in my graph json to show the user which criteria have failed for them. I can see that the response does contain a trace field. Here's the response struct for your reference : 

type EvaluationResponse struct {
	Performance string           `json:"performance"`
	Result      json.RawMessage  `json:"result"`
	Trace       *json.RawMessage `json:"trace"`
}

However, the response that I am receiving contains nil value for the Trace field. It would be really helpful if someone can guide me on how to get the same in the response itself.

Thanks

**Comment by divyeshk31** [comment]
@ivanmiletic pls help

**Comment by stefan-gorules** [comment]
Please refer to: 

https://github.com/gorules/zen-go/blob/adbba1f342a10587affb701d6316cee89fbdf40a/engine_test.go#L123-L145

**Reply by aayushbhaglal** [reply]
Hi Stefan,

Thanks for the guidance. I tried out your approach by doing something like this : 
`decisionResponse, err := l.Decision[*lead.ProductType].EvaluateWithOpts(input, zen.EvaluationOptions{
		Trace:    true,
		MaxDepth: 10,
	})`

However the response that I am receiving in trace doesn't contain much to assess as to which nodes failed so that I can give the user a feedback. Here's the trace that I received : 

`{
  "640faf70-e9e1-4459-be2f-abd4ca18da74": {
    "input": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "output": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "name": "age check",
    "id": "640faf70-e9e1-4459-be2f-abd4ca18da74",
    "performance": "5.5µs",
    "traceData": null
  },
  "5acf288c-223a-4b7a-a088-cc6a67969bb5": {
    "input": null,
    "output": null,
    "name": "request",
    "id": "5acf288c-223a-4b7a-a088-cc6a67969bb5",
    "performance": null,
    "traceData": null
  },
  "989ebc2e-94ce-4eb7-a174-e13a512c2be7": {
    "input": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "output": {},
    "name": "Not Eligible",
    "id": "989ebc2e-94ce-4eb7-a174-e13a512c2be7",
    "performance": "24.625µs",
    "traceData": {}
  },
  "9762a89c-ec2b-45b8-83bf-7af3689e046a": {
    "input": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "output": {},
    "name": "Final Evaluation",
    "id": "9762a89c-ec2b-45b8-83bf-7af3689e046a",
    "performance": "29.459µs",
    "traceData": {}
  },
  "0ab7d317-0fcb-4e84-9f08-7a22817a2fb6": {
    "input": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "output": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "name": "Constitution Check",
    "id": "0ab7d317-0fcb-4e84-9f08-7a22817a2fb6",
    "performance": "4.125µs",
    "traceData": null
  },
  "2df654d3-f6a2-471f-bad0-ade5cc1c4294": {
    "input": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "output": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "name": "Vintage check",
    "id": "2df654d3-f6a2-471f-bad0-ade5cc1c4294",
    "performance": "4.709µs",
    "traceData": null
  },
  "154c8048-e62f-4d98-9f36-8ce4e12f2474": {
    "input": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "output": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "name": "Credit score check",
    "id": "154c8048-e62f-4d98-9f36-8ce4e12f2474",
    "performance": "3.916µs",
    "traceData": null
  },
  "987318c2-a213-4119-b123-504ba77773d3": {
    "input": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "output": {
      "request": {
        "age": 21,
        "cscore": 764,
        "employmentType": "Salaried",
        "loanAmount": 500000,
        "monthlyIncome": 250000,
        "pincode": 1234,
        "turnover": 0,
        "vintage": 0
      }
    },
    "name": "Turnover Check",
    "id": "987318c2-a213-4119-b123-504ba77773d3",
    "performance": "4.667µs",
    "traceData": null
  },
  "834668d4-bccd-4153-b9d6-8cf26c7b5b0c": {
    "input": null,
    "output": null,
    "name": "response",
    "id": "834668d4-bccd-4153-b9d6-8cf26c7b5b0c",
    "performance": null,
    "traceData": null
  }
}`

**Reply by aayushbhaglal** [reply]
@stefan-gorules

**Reply by stefan-steadfast** [reply]
Hi @aayushbhaglal, the trace data is useful for tracing how the decision was made, not necessarily for failure. Failures are thrown as an error instead.

What kind of error are you looking to provide to the user? Error while they are building decision, or while they are interacting with some API?

**Reply by aayushbhaglal** [reply]
Hi @stefan-steadfast, the trace data I am receiving for the decision doesn't contain any information on individual nodes which failed. All I am receiving is the final output by the Evaluate function which tells me whether the user is finally eligible or ineligible. i want to show the user which nodes returned false in case they are ineligible.

**Reply by stefan-gorules** [reply]
Based on the trace data, nothing has failed seemingly. It could be that no decision has been hit, but there aren't any errors per se.

Empty trace data on decision tables for example indicate that no rule has been hit. We will be improving trace data in the very near future. We can prioritise features like this for business and enterprise users, and provide better usage support. Feel free to reach us at hi@gorules.io.

**Reply by aayushbhaglal** [reply]
Hi @stefan-gorules ,

I have further investigated this and found a potential bug. The trace data is accurate in case of decision table, but in case of other functions like switch, the input itself is returned in the trace output. I am attaching screenshots for both the cases for your reference.
<img width="706" alt="Screenshot 2024-10-22 at 2 38 16 PM" src="https://github.com/user-attachments/assets/6476157d-877f-43a8-b15f-23eba472a034">
<img width="644" alt="Screenshot 2024-10-22 at 2 39 00 PM" src="https://github.com/user-attachments/assets/45892e52-6ba0-403c-a133-ee21be947344">
